### DISCUSSION TITLE
HASS diagnostic performance counters to enable quicker troubleshooting

### DISCUSSION THREAD

**Discussion by PeteRager** [home-assistant/architecture#866] [Entity Models]
There is a need to be able to assess the performance of a running HASS system.  The current tools measuring CPU and Memory while useful are often not actionable.  There are numerous examples in the forums of root causes being entities updating very frequently.  Unfortunately, it takes significant effort to capture log files and expertise to identify the entities and create guidance.

The vision would be to have the equivalent of a Unix Top view that shows the most active entities, events, etc in the system.  This would allow the end user to immediately identify the most chatty entities in the system and take action.  HASS could also provide user notifications when these exceed a threshold.

I’m willing to do the work to get this started.  Purpose of this discussion is to determine the approach and appetite of potentially making changes to the core.

While this functionally could be done as an integration by subscribing to events, it would be suboptimal from a performance perspective.  It is desirable for performance counter collection have a very minimal impact on the system.

The first step could be to create counters for each entity that increment when the entity state is updated or changed. This would allow identification of the most frequently updating entities.   This would not allow identification of bursts (e.g entity updates 100x in a minute, then no updates for an hour.). However it is a first step and is 80% use case.

Right now there are last_updated and last_changed timestamps on the entity.  An approach could be add two counters update_count and changed_count.  They initialize to zero and then increment in the branches that update the timestamps.

**Comment by allenporter** [comment]
I have some more general thoughts beyond performance diagnostics. I have added some counters to existing diagnostics to help troubleshoot integration behavior, though not specifically for performance. Adding counters to diagnostics is straight forward and useful and is possible today, so this makes me think about these questions:
- Is it useful to define some patterns or helper libraries for implementing counters in diagnostics?
- Are there a set of additional default counters related to performance that might be interesting to define and include?

To give an example of a concrete use case I have here already, I use counters to understand what is happening in an integration behind the scenes to answer questions that might be a bit too spammy to log otherwise (e.g.  did the user not see the camera motion event because the subscriber didn't start? the message  was never received? or because the media url expired?)

I end up having a simple `Counter` like class that gets dumped to a format that can be included in the `diagnostics` integration which can  track events, or latency of events [example for nest](https://github.com/allenporter/python-google-nest-sdm/blob/main/google_nest_sdm/diagnostics.py#L12).  

The resulting diagnostics for a camera look something like this:

```
        "event_media": {
          "event": 2056,
...
          "event.new": 735,
          "event.fetch": 2056,
          "fetch_image": 2056,
          "fetch_image.skip": 334,
          "event.update": 1321,
          "sdm.devices.events.CameraMotion.Motion_count": 1823,
          "sdm.devices.events.CameraMotion.Motion_sum": 903686,
          "fetch_image.save": 2057,
...
          "event.fetch_error": 6,
        }
```

This tells me that this device has received 2056 camera events, and 1823 of them were motion events.  Camera motion events for this device take an average of 495ms to be delivered.  (latency sum divided by count, following how you might do this with prometheus).

**Reply by PeteRager** [reply]
Those are good questions.  There is diagnostic/performance information outside of the core to consider also and having a consistent efficient way to do it would be helpful.  Diagnostics can be at the integration and/or device level.  Yours is a great example of device diagnostics that make it easier and quicker to diagnose problems.

Today I had an issue with an automation.  Eventually deep down in a zwavejs log file I found that the command had timed out.  Would have been a lot easier to be in HASS, click on the device and see that the timeout counter for the device had increased.  So I agree getting better diags into complex integrations - zwave, MQTT, etc would be helpful.

In the Lennox integration I manage.  I added device diagnostics as attributes inside an entity that reports the connection state.  With this most of my users are self sufficient in diagnosing connectivity issues and it’s easy to create a HASS dashboard. 

I did prototype a quick change to track state updates and make the attr available in Lovelace.  I am going to look at Prometheus and there counter libraries.  Certainly a consideration is adding size to the state class and increasing memory usage.

![402761F2-B167-4CED-90D8-F4FE92DDEB05](https://user-images.githubusercontent.com/76050312/216460462-825bea0c-6ae9-4eaa-8713-70e3439bc666.jpeg)

**Comment by PeteRager** [comment]
A few weeks ago, I noticed an issue with a zwave command not always working and errors not being reported in addition to no diagnostics in HASS to help.   I’ve also had issues in the past with NEST setpoints not taking and having to issue retries.  Also, in the forums there are many similar issues and people having to code retry loops in there automations.  Since these a all point solutions, and I did not want to code yet another retry loop…I’ve decided to create a better way. 

As a result, I’ve shifted my current focus on diagnostics around service calls.  To be able to answer the following questions:
- How many times is the service call happening per entity?
- How long does it take for the call to execute and the expected response to occur?
- How many times does the expected response not occur within a configured timeout?
- How many times is the service call retried?
- How many times does it fail?

The first implementation I’ve created uses shell scripts to generate HASS scripts.  The next implementation is as a HASS integration.

I can certainly package this as a custom integration.  My question is would something like this be good for inclusion in HASS? And if so, there are a number of design choices that we may want agreement on.

here are some screenshots with data 

Zwave Set temperature. Typically fast, 150ms, but there was a 12 hour period of time where everything was much slower. Preliminary analysis showed that the delay was not in zwave but perhaps something stalling HASS event loop for 500ms?

![49CE6F46-080F-4128-8A7F-5A56FB268DBE](https://user-images.githubusercontent.com/76050312/219957365-1b9084d6-3c2f-43ee-bc30-929834b52c58.jpeg)

nest set temperature latency.  Typically occurs within 2-5 seconds, there are sometimes that the change takes longer and needs to be retried.

![C733AAAC-1965-4002-8932-094E79113BDF](https://user-images.githubusercontent.com/76050312/219957367-83305117-f88d-4b41-a63f-351a80b5f9a7.jpeg)


![A28B6E02-6E02-42BA-BDC2-C3088B930CB6](https://user-images.githubusercontent.com/76050312/219957368-efeff0c6-3141-4006-af4d-cc2e191186cf.jpeg)




https://github.com/PeteRager/hass_rs

**Reply by allenporter** [reply]
You should see similar data in the nest diagnostics with the counters I mentioned. It won't give you history or graph, but has aggregate stats like average latency or total failure rate. Could be interesting to compare to what you have collected.

Generally I'm suggesting exploring this approach as it already exists, is baked into our developer bug report flow ,(e.g. users give it to you), and it is hidden from most users, but advanced users could dig in if needed.
